### PR TITLE
direct transfer state change

### PR DESCRIPTION
**Pull request by hackaugusto**
#588

### CODE REVIEW

File: raiden/transfer/events.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from raiden.transfer.architecture import Event
+# pylint: disable=too-many-arguments,too-few-public-methods
+
+
+class EventTransferSentSuccess(Event):
+    """ Event emitted by the initiator when a transfer is considered sucessful.
+
+    A transfer is considered sucessful when the initiator's payee hop sends the
+    reveal secret message, assuming that each hop in the mediator chain has
+    also learned the secret and unlock/withdraw it's token.
```
Comment (LefterisJP) [inline raiden/transfer/events.py]: `unlock/withdraw it's` -> `unlocks/withdraws its`

File: raiden/transfer/events.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      assume byzantine behavior the message is considered delivered without an
+      acknowledgement.
+    - The transfer is considered sucessful because of the above.
+    - The reveal secret message was not delivered because of actual network
+      problems.
+    - The lock expires and an EventUnlockFailed follows, contradicting the
+      EventTransferSentSuccess.
+
+    Note:
+        Mediators cannot use this event, since a unlock may be locally
```
Comment (LefterisJP) [inline raiden/transfer/events.py]: `a unlock` -> `an unlock`

File: raiden/transfer/mediated_transfer/events.py [hunk 3]
```diff
@@ -41,17 +41,33 @@ def __init__(
 
 
 class SendRevealSecret(Event):
-    """ Event used to send a reveal the secret to another node, not the same as
-    a balance-proof.
+    """ Sends a RevealSecret to another node.
 
-    Used by payees: The target and mediator nodes.
+    This event is used once the secret is know locally and an action must be
```
Comment (LefterisJP) [inline raiden/transfer/mediated_transfer/events.py]: `know`->`known`

File: raiden/transfer/mediated_transfer/events.py [hunk 4]
```diff
@@ -41,17 +41,33 @@ def __init__(
...
+    performed on the receiver:
+
+        - For receivers in the payee role, it informs the node that the lock has
+        been released and the token can be withdraw, either on-chain or
```
Comment (LefterisJP) [inline raiden/transfer/mediated_transfer/events.py]: `withdraw` -> `withdrawn`

File: raiden/transfer/mediated_transfer/events.py [hunk 5]
```diff
@@ -41,17 +41,33 @@ def __init__(
...
+
+    For any mediated transfer:
+
+        - The initiator will only perform the payer role.
+        - The target will only perform the payee role.
+        - The mediators will have `n` channels at the payee role and `n` at the
+          payer role, where `n` is equal to `1 + number_of_refunds`.
 
     Note:
         The payee must only update it's local balance once the payer sends an
```
Comment (LefterisJP) [inline raiden/transfer/mediated_transfer/events.py]: `it's` -> `its`

File: raiden/transfer/mediated_transfer/initiator.py [hunk 6]
```diff
@@ -209,7 +229,8 @@ def handle_secretrequest(state, state_change):
 
     if valid_secretrequest:
         # Reveal the secret to the target node and wait for it's confirmation,
```
Comment (LefterisJP) [inline raiden/transfer/mediated_transfer/initiator.py]: `it's` -> `its`

File: raiden/transfer/state_change.py [hunk 7]
```diff
@@ -43,3 +43,29 @@ class ActionCancelTransfer(StateChange):
 
     def __init__(self, identifier):
         self.identifier = identifier
+
+
+class ActionTransferDirect(StateChange):
+    def __init__(self,
```
Comment (LefterisJP) [inline raiden/transfer/state_change.py]: Can you put all the arguments in the next line, like we have in the style guide?

File: raiden/raiden_service.py [hunk 8]
```diff
@@ -1056,8 +1081,22 @@ def message_directtransfer(self, message):
                 )
             )
 
+        amount = channel.partner_state.transferred_amount - message.transferred_amount
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: yes, you're right, fixed and added tests for it

File: raiden/tests/unit/transfer/mediated_transfer/test_initiator.py [hunk 9]
```diff
@@ -127,9 +130,9 @@ def test_next_route():
     assert state.route == routes[0], 'a initialized state must use the first valid route'
 
     assert state.routes.available_routes == routes[1:]
-    assert len(state.routes.ignored_routes) == 0
-    assert len(state.routes.refunded_routes) == 0
-    assert len(state.routes.canceled_routes) == 0
+    assert not state.routes.ignored_routes
```
Comment (hackaugusto) [inline raiden/tests/unit/transfer/mediated_transfer/test_initiator.py]: if you want some sort of 'type safety' the best way is to use `isinstance`, the ignored_routes could also be a set or a dict and `len` would work. In general the preferred approach is the most readable one, and for checking if a collection has elements `assert collection` is the preferred way.

Comment (LefterisJP) [inline raiden/tests/unit/transfer/mediated_transfer/test_initiator.py]: I still find the `len(collection) == 0` more readable and explicit but it seems the pythonic way to do it is with the `not collection` so I am fine.

**Pull request merged**